### PR TITLE
[Drill-3710] New option for the IN LIST size to convert into join

### PR DESCRIPTION
**Pull request by gparai**
Add option planner.in_subquery_threshold to control the size of the IN list for converting to join

### CODE REVIEW

File: exec/java-exec/src/test/java/org/apache/drill/TestPartitionFilter.java [hunk 1]
```diff
@@ -376,4 +376,14 @@ public void testPartitionFilterWithLike() throws Exception {
     testIncludeFilter(query4, 4, "Filter", 16);
   }
 
+  @Test //DRILL-3710 Partition pruning should occur with varying IN-LIST size
+  public void testPartitionFilterWithInSubquery() throws Exception {
+    String query = String.format("select * from dfs_test.`%s/multilevel/parquet` where cast (dir0 as int) IN (1994, 1994, 1994, 1994, 1994, 1994)", TEST_RES_PATH);
+    /* In list size exceeds threshold - no partition pruning since predicate converted to join */
+    test("alter session set `planner.in_subquery_threshold` = 2");
```
Comment (amansinha100) [inline exec/java-exec/src/test/java/org/apache/drill/TestPartitionFilter.java]: Not sure if it is necessary to check the no-partition-pruning case.  Basically, the goal of the test is to see if partition pruning works with large IN lists.

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/PlannerSettings.java [hunk 2]
```diff
@@ -83,6 +83,8 @@
 
   public static final String TYPE_INFERENCE_KEY = "planner.enable_type_inference";
   public static final BooleanValidator TYPE_INFERENCE = new BooleanValidator(TYPE_INFERENCE_KEY, true);
+  public static final OptionValidator IN_SUBQUERY_THRESHOLD =
+          new PositiveLongValidator("planner.in_subquery_threshold", Integer.MAX_VALUE, 20);
```
Comment (amansinha100) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/PlannerSettings.java]: you should add a comment here for the default value of 20 is chosen to match the default Calcite value.

**Comment by amansinha100** [comment]
couple of minor comments.  Overall +1

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/PlannerSettings.java [hunk 3]
```diff
@@ -216,6 +218,10 @@ public boolean isTypeInferenceEnabled() {
     return options.getOption(TYPE_INFERENCE);
   }
 
+  public long getInSubqueryThreshold() {
+    return options.getOption(IN_SUBQUERY_THRESHOLD.getOptionName()).num_val;
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/PlannerSettings.java]: Change the declaration to: `public static final LongValidator IN_SUBQUERY_THRESHOLD ...`

and here, `return options.getOption(IN_SUBQUERY_THRESHOLD;`

File: exec/java-exec/src/test/java/org/apache/drill/TestPartitionFilter.java (hunk 1 above)
Comment (gparai) [inline exec/java-exec/src/test/java/org/apache/drill/TestPartitionFilter.java]: A bug could cause us to not obey the option at all i.e. we always do partition pruning regardless of the option setting. This unit test checks we do obey the option.

**Comment by amansinha100** [comment]
+1

**Pull request closed without merging**